File: task-2-mobile-automation-python/src/pages/article_page.py
from __future__ import annotations

from selenium.webdriver.common.by import By
from appium.webdriver.common.appiumby import AppiumBy

from src.pages.base_page import BasePage


class ArticlePage(BasePage):
    ARTICLE_TITLE = (By.ID, "org.wikipedia:id/view_page_title_text")
    # The bookmark/save icon in the article's bottom toolbar. Tapping it the
    # first time saves to the default list; the "add to a specific list"
    # flow (SaveToReadingListDialog) is what this test suite exercises.
    SAVE_BUTTON = (AppiumBy.ACCESSIBILITY_ID, "Save page")
    SAVED_BUTTON_STATE = (AppiumBy.ACCESSIBILITY_ID, "Saved")

    def get_title(self) -> str:
        return self.wait_visible(self.ARTICLE_TITLE).text

    def open_add_to_reading_list(self):
        from src.pages.save_to_reading_list_dialog import SaveToReadingListDialog

        self.click(self.SAVE_BUTTON)
        return SaveToReadingListDialog(self.driver)

    def is_saved(self) -> bool:
        return self.is_present(self.SAVED_BUTTON_STATE)

File: task-2-mobile-automation/src/main/java/com/wikipedia/automation/pages/ArticlePage.java
package com.wikipedia.automation.pages;

import io.appium.java_client.AppiumDriver;
import org.openqa.selenium.By;

public class ArticlePage extends BasePage {

    private final By articleTitle = By.id("org.wikipedia:id/view_page_title_text");
    // The bookmark/save icon in the article's bottom toolbar. Tapping it the
    // first time saves to the default list; tapping again on an already
    // saved article opens/removes it. Long variants of the Wikipedia app
    // route "add to a specific list" through the overflow/bookmark-long-press
    // flow captured in SaveToReadingListDialog.
    private final By saveButton = By.AccessibilityId("Save page");
    private final By savedButtonState = By.AccessibilityId("Saved");

    public ArticlePage(AppiumDriver driver) {
        super(driver);
    }

    public String getTitle() {
        return waitVisible(articleTitle).getText();
    }

    /** Opens the "add to reading list" bottom sheet (long-press equivalent / tap-and-hold on save icon). */
    public SaveToReadingListDialog openAddToReadingList() {
        waitClickable(saveButton).click();
        return new SaveToReadingListDialog(driver);
    }

    public boolean isSaved() {
        return isPresent(savedButtonState);
    }
}
